makelib/imagelight.mk

```diff
@@ -0,0 +1,214 @@
+	@$(OK) cleaning images for $(BUILD_REGISTRY)
+
+img.done:
+	@rm -fr $(IMAGE_TEMP_DIR)
+
+img.cache:
+	@for i in $(CACHE_IMAGES); do \
+		IMGID=$$(docker images -q $$i); \
+		if [ -n "$$IMGID" ]; then \
+			echo === caching image $$i; \
+			CACHE_IMAGE=$(CACHE_REGISTRY)/$${i#*/}; \
+			docker tag $$i $${CACHE_IMAGE}:$(CACHE_TAG); \
+			for r in $$(docker images --format "{{.ID}}#{{.Repository}}:{{.Tag}}" | grep $$IMGID | grep $(CACHE_REGISTRY)/ | grep -v $${CACHE_IMAGE}:$(CACHE_TAG)); do \
+				docker rmi $${r#*#} > /dev/null 2>&1 || true; \
+			done; \
+		fi; \
+	done
+
+# prune removes old cached images
+img.prune:
+	@$(INFO) pruning images older than $(PRUNE_HOURS) keeping a minimum of $(PRUNE_KEEP) images
+	@EXPIRED=$$(docker images --format "{{.Tag}}#{{.Repository}}:{{.Tag}}" \
+		| grep -E '$(CACHE_REGISTRY)/' \
+		| sort -r \
+		| awk -v i=0 -v cd="$(CACHE_PRUNE_DATE)" -F  "#" '{if ($$1 <= cd && i >= $(PRUNE_KEEP)) print $$2; i++ }') &&\
+	for i in $$EXPIRED; do \
+		echo removing expired cache image $$i; \
+		[ $(PRUNE_DRYRUN) = 1 ] || docker rmi $$i > /dev/null 2>&1 || true; \
+	done
+	@for i in $$(docker images -q -f dangling=true); do \
+		echo removing dangling image $$i; \
+		docker rmi $$i > /dev/null 2>&1 || true; \
+	done
+	@$(OK) pruning
+
+debug.nuke:
+	@for c in $$(docker ps -a -q --no-trunc); do \
+		if [ "$$c" != "$(SELF_CID)" ]; then \
+			echo stopping and removing container $${c}; \
+			docker stop $${c}; \
+			docker rm $${c}; \
+		fi; \
+	done
+	@for i in $$(docker images -q); do \
+		echo removing image $$i; \
+		docker rmi -f $$i > /dev/null 2>&1; \
+	done
+
+# 1: registry 2: image
+define repo.targets
+img.release.publish.$(1).$(2):
+	@$(MAKE) -C $(IMAGE_DIR)/$(2) IMAGE_PLATFORMS=$(IMAGE_PLATFORMS) IMAGE=$(1)/$(2):$(VERSION) img.publish
+img.release.publish: img.release.publish.$(1).$(2)
+
+img.release.promote.$(1).$(2):
+	@$(MAKE) -C $(IMAGE_DIR)/$(2) TO_IMAGE=$(1)/$(2):$(CHANNEL) FROM_IMAGE=$(1)/$(2):$(VERSION) img.promote
+	@[ "$(CHANNEL)" = "master" ] || $(MAKE) -C $(IMAGE_DIR)/$(2) TO_IMAGE=$(1)/$(2):$(VERSION)-$(CHANNEL) FROM_IMAGE=$(1)/$(2):$(VERSION) img.promote
+img.release.promote: img.release.promote.$(1).$(2)
+
+img.release.clean.$(1).$(2):
+	@[ -z "$$$$(docker images -q $(1)/$(2):$(VERSION))" ] || docker rmi $(1)/$(2):$(VERSION)
+	@[ -z "$$$$(docker images -q $(1)/$(2):$(VERSION)-$(CHANNEL))" ] || docker rmi $(1)/$(2):$(VERSION)-$(CHANNEL)
+	@[ -z "$$$$(docker images -q $(1)/$(2):$(CHANNEL))" ] || docker rmi $(1)/$(2):$(CHANNEL)
+img.release.clean: img.release.clean.$(1).$(2)
+endef
+$(foreach r,$(REGISTRY_ORGS), $(foreach i,$(IMAGES),$(eval $(call repo.targets,$(r),$(i)))))
+
+# ====================================================================================
+# Common Targets
+
+do.build.image.%:
+	@$(MAKE) -C $(IMAGE_DIR)/$* IMAGE_PLATFORMS=$(IMAGE_PLATFORM) IMAGE=$(BUILD_REGISTRY)/$*-$(ARCH) img.build
+do.build.images: $(foreach i,$(IMAGES), do.build.image.$(i))
```

Comment:
So this piece may answer part of my "how are we utilizing the cache" question. It looks like we're narrowing down to one platform at a time to build the images. Given that, I am curious if `buildx` will experience the same problem we saw with `docker build` and the mutli-arch static container when running parallel image builds for each target platform.

Reply:
@tnthornton this does call `img.build` individually for each platform for each image 👍🏻 I am fairly certain that we will not experience this race condition with `buildx` as the underlying builders are intended to be able to [handle parallel requests](https://github.com/docker/buildx/issues/344#issuecomment-670152753). This is essentially [how `docker buildx bake` works](https://github.com/docker/buildx/blob/master/docs/reference/buildx_bake.md). This is aligned with the tests I have ran with parallelism as well.

One other thing to note here is that we do run individual builds to load into the daemon for local dev and CI, but when we push we use `--platforms` and `--push` to build a single mutli-arch image (see example Makefile in PR body).